File: backend/app/vertex.py
from __future__ import annotations

import time
from typing import Any

from google.cloud import aiplatform_v1

from .config import Settings


class VertexClient:
    def __init__(self, settings: Settings) -> None:
        self.settings = settings

    def predict(self, prompt: str, overrides: dict[str, Any] | None = None) -> tuple[str, int]:
        start = time.time()

        config = _merge_overrides(self.settings, overrides or {})
        instance = _render_template(config.instance_template, prompt)
        parameters = _render_template(config.parameters_template, prompt)

        client = aiplatform_v1.PredictionServiceClient(
            client_options={"api_endpoint": f"{config.location}-aiplatform.googleapis.com"}
        )
        endpoint = client.endpoint_path(config.project_id, config.location, config.endpoint_id)

        response = client.predict(
            endpoint=endpoint,
            instances=[instance],
            parameters=parameters,
        )

        text = _extract_prediction_text(response.predictions)
        latency_ms = int((time.time() - start) * 1000)
        return text, latency_ms


def _merge_overrides(settings: Settings, overrides: dict[str, Any]) -> _ResolvedConfig:
    project_id = overrides.get("project_id") or settings.vertex_project_id
    endpoint_id = overrides.get("endpoint_id") or settings.vertex_endpoint_id
    location = overrides.get("location") or settings.vertex_location

    if not project_id or not endpoint_id:
        raise ValueError("Vertex project_id and endpoint_id are required")

    if "instance_template" in overrides:
        instance_template = overrides["instance_template"]
    else:
        instance_template = settings.parse_json_template(settings.vertex_instance_template)

    if "parameters_template" in overrides:
        parameters_template = overrides["parameters_template"]
    else:
        parameters_template = settings.parse_json_template(settings.vertex_parameters_template)

    return _ResolvedConfig(
        project_id=project_id,
        endpoint_id=endpoint_id,
        location=location,
        instance_template=instance_template,
        parameters_template=parameters_template,
    )


def _render_template(template: dict[str, Any], prompt: str) -> dict[str, Any]:
    def render(value: Any) -> Any:
        if isinstance(value, str):
            return value.replace("{prompt}", prompt)
        if isinstance(value, dict):
            return {key: render(val) for key, val in value.items()}
        if isinstance(value, list):
            return [render(item) for item in value]
        return value

    return render(template)


def _extract_prediction_text(predictions: Any) -> str:
    if not predictions:
        return ""

    first = predictions[0]
    if isinstance(first, str):
        return first

    if isinstance(first, dict):
        for key in ("content", "text", "output", "prediction", "generated_text", "response"):
            if key in first and isinstance(first[key], str):
                return first[key]
        if "candidates" in first and first["candidates"]:
            candidate = first["candidates"][0]
            if isinstance(candidate, dict):
                for key in ("content", "text", "output"):
                    if key in candidate and isinstance(candidate[key], str):
                        return candidate[key]
            if isinstance(candidate, str):
                return candidate

    return str(first)


class _ResolvedConfig:
    def __init__(
        self,
        project_id: str,
        endpoint_id: str,
        location: str,
        instance_template: dict[str, Any],
        parameters_template: dict[str, Any],
    ) -> None:
        self.project_id = project_id
        self.endpoint_id = endpoint_id
        self.location = location
        self.instance_template = instance_template
        self.parameters_template = parameters_template

File: backend/src/vertex.js
"use strict";

const { performance } = require("node:perf_hooks");
const { GoogleAuth } = require("google-auth-library");
const { parseJsonTemplate } = require("./config");
const { toLogJson } = require("./logging");

class VertexClient {
  constructor(settings) {
    this.settings = settings;
    this.auth = new GoogleAuth({
      scopes: ["https://www.googleapis.com/auth/cloud-platform"]
    });
  }

  async predict(prompt, overrides = {}) {
    const start = performance.now();
    const config = mergeOverrides(this.settings, overrides);
    const instance = renderTemplate(config.instanceTemplate, prompt);
    const parameters = renderTemplate(config.parametersTemplate, prompt);
    const endpoint = `projects/${config.projectId}/locations/${config.location}/endpoints/${config.endpointId}`;

    console.info(
      "VERTEX REQUEST endpoint=%s instance=%s parameters=%s",
      endpoint,
      toLogJson(instance),
      toLogJson(parameters)
    );

    const token = await this.getAccessToken(overrides);
    const url = `https://${config.location}-aiplatform.googleapis.com/v1/${endpoint}:predict`;

    const response = await fetchWithTimeout(
      url,
      {
        method: "POST",
        headers: {
          authorization: `Bearer ${token}`,
          "content-type": "application/json"
        },
        body: JSON.stringify({
          instances: [instance],
          parameters
        })
      },
      this.settings.requestTimeoutSec * 1000
    );

    const body = await safeJson(response);
    if (!response.ok) {
      throw new Error(formatError("Vertex", response.status, body));
    }

    const predictions = Array.isArray(body?.predictions) ? body.predictions : [];
    console.info("VERTEX RESPONSE predictions=%s", toLogJson(predictions));

    const text = extractPredictionText(predictions);
    const latencyMs = Math.round(performance.now() - start);
    return [text, latencyMs];
  }

  async getAccessToken(overrides) {
    if (overrides.access_token) {
      return overrides.access_token;
    }
    const client = await this.auth.getClient();
    const tokenResult = await client.getAccessToken();
    const token = typeof tokenResult === "string" ? tokenResult : tokenResult?.token;
    if (!token) {
      throw new Error("Failed to acquire Google access token");
    }
    return token;
  }
}

function mergeOverrides(settings, overrides) {
  const projectId = overrides.project_id || settings.vertexProjectId;
  const endpointId = overrides.endpoint_id || settings.vertexEndpointId;
  const location = overrides.location || settings.vertexLocation;

  if (!projectId || !endpointId) {
    throw new Error("Vertex project_id and endpoint_id are required");
  }

  const instanceTemplate = isObject(overrides.instance_template)
    ? overrides.instance_template
    : parseJsonTemplate(settings.vertexInstanceTemplate, "BESCO_VERTEX_INSTANCE_TEMPLATE");

  const parametersTemplate = isObject(overrides.parameters_template)
    ? overrides.parameters_template
    : parseJsonTemplate(settings.vertexParametersTemplate, "BESCO_VERTEX_PARAMETERS_TEMPLATE");

  return {
    projectId,
    endpointId,
    location,
    instanceTemplate,
    parametersTemplate
  };
}

function renderTemplate(template, prompt) {
  if (Array.isArray(template)) {
    return template.map((item) => renderTemplate(item, prompt));
  }
  if (template && typeof template === "object") {
    const out = {};
    for (const [key, value] of Object.entries(template)) {
      out[key] = renderTemplate(value, prompt);
    }
    return out;
  }
  if (typeof template === "string") {
    return template.replaceAll("{prompt}", prompt);
  }
  return template;
}

function extractPredictionText(predictions) {
  if (!predictions.length) {
    return "";
  }
  const first = predictions[0];
  if (typeof first === "string") {
    return first;
  }
  if (first && typeof first === "object") {
    for (const key of ["content", "text", "output", "prediction", "generated_text", "response"]) {
      if (typeof first[key] === "string") {
        return first[key];
      }
    }
    if (Array.isArray(first.candidates) && first.candidates.length) {
      const candidate = first.candidates[0];
      if (typeof candidate === "string") {
        return candidate;
      }
      if (candidate && typeof candidate === "object") {
        for (const key of ["content", "text", "output"]) {
          if (typeof candidate[key] === "string") {
            return candidate[key];
          }
        }
      }
    }
  }
  return String(first);
}

function formatError(provider, status, body) {
  const error = body?.error ?? body;
  const message = typeof error === "object" ? error?.message || JSON.stringify(error) : String(error);
  return `${provider} request failed (${status}): ${message}`;
}

function safeJson(response) {
  return response.text().then((text) => {
    if (!text) {
      return {};
    }
    try {
      return JSON.parse(text);
    } catch {
      return { text };
    }
  });
}

async function fetchWithTimeout(url, options, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, {
      ...options,
      signal: controller.signal
    });
  } catch (error) {
    if (error && error.name === "AbortError") {
      throw new Error("Request timed out");
    }
    throw error;
  } finally {
    clearTimeout(timer);
  }
}

function isObject(value) {
  return value && typeof value === "object" && !Array.isArray(value);
}

module.exports = {
  VertexClient
};
